rest_framework_json_api/backends.py

```diff
@@ -0,0 +1,148 @@
+    search_param = api_settings.SEARCH_PARAM
+    ordering_param = 'sort'
+
+    def __init__(self):
+        self.filter_keys = []
+
+
+class JsonApiQueryValidationFilter(JsonApiFilterMixin, BaseFilterBackend):
+    """
+    A backend filter that validates query parameters for jsonapi spec conformance and raises a 400 error
+    rather than silently ignoring unknown parameters or incorrect usage.
+
+    set `allow_duplicate_filters = True` if you are OK with the same filter being repeated.
+
+    TODO: For jsonapi error object compliance, must set jsonapi errors "parameter" for the ValidationError.
+          This requires extending DRF/DJA Exceptions.
+    """
+    allow_duplicated_filters = False
+
+    def validate_query_params(self, request):
+        """
+        Validate that query params are in the list of valid `jsonapi_query_keywords`
+        Raises ValidationError if not.
+        """
+        for qp in request.query_params.keys():
+            bracket = qp.find('[')
+            if bracket >= 0:
+                if qp[-1] != ']':
+                    raise ValidationError(
+                        'invalid query parameter: {}'.format(qp))
+                keyword = qp[:bracket]
+            else:
+                keyword = qp
+            if keyword not in self.jsonapi_query_keywords:
+                raise ValidationError(
+                    'invalid query parameter: {}'.format(keyword))
+            if not self.allow_duplicated_filters and len(
+                    request.query_params.getlist(qp)) > 1:
+                raise ValidationError(
+                    'repeated query parameter not allowed: {}'.format(qp))
+
+    def filter_queryset(self, request, queryset, view):
+        self.validate_query_params(request)
+        return queryset
+
+
+class JsonApiOrderingFilter(JsonApiFilterMixin, OrderingFilter):
+    """
+    The standard rest_framework.filters.OrderingFilter works mostly fine as is, but with .ordering_param = 'sort'.
+
+    This implements http://jsonapi.org/format/#fetching-sorting and raises 400 if any sort field is invalid.
+    """
+    ignore_bad_sort_fields = False
+
+    def remove_invalid_fields(self, queryset, fields, view, request):
+        """
+        override remove_invalid_fields to raise a 400 exception instead of silently removing them.
+        """
+        valid_fields = [item[0] for item in self.get_valid_fields(queryset, view, {'request': request})]
+        bad_terms = [term for term in fields if term.lstrip('-') not in valid_fields and ORDER_PATTERN.match(term)]
+        if bad_terms and not self.ignore_bad_sort_fields:
+            raise ValidationError(
+                'invalid sort parameter{}: {}'.format(('s' if len(bad_terms) > 1 else ''), ','.join(bad_terms)))
+        return super(JsonApiOrderingFilter, self).remove_invalid_fields(queryset, fields, view, request)
+
+
+class JsonApiSearchFilter(JsonApiFilterMixin, SearchFilter):
```

Comment:
If we just include `JSONAPISearchFilter` in `DEFAULT_FILTER_BACKENDS` or the `filter_backends` attribute then the default value of `filter[search]`, `filter[all]` or whatever is predefined. Otherwise, including plain `SearchFilter` will require one to also set settings.JSON_API_SEARCH_PARAM or `search_param` attribute in order to comply with the requirement to use `filter`. So it's less typing to just use the JSONAPI-flavored class even if it doesn't do much.

Regarding the name of the filter, I like `all` (adverb) because it implies search *all* the (searchable) fields whereas `search` feels more like a verb usage to me which is anti-REST. But you can use one and I can use the other as it's configurable.

Reply:
I kind of think `filter[all]` is misleading as it implies that it takes into account all fields for filtering. This is actually not true as it only search in the defined `search_fields`. `filter[search]` might also be a bit odd though I agree.

Even though everyone can define the name as they like if we add it to DJA we have to come up with a recommendation. But this doesn't seem to be easy as the json api is silent on it and we are just two users and have two different opinions already :smile: 

While thinking it seems that adding a `JSONAPISearchFilter` to DJA doesn't add any value as someone can already use the DRF `SearchFilter` today without this change and define whatever parameter he likes.

Also in DJA we don't have to reimplement all features which are in DRF but only implement what is different according to the json api specification. 

For this reasons I think we shouldn't provide a `JSONAPISearchFilter`. We might add a reference to the DRF documentation that there are other filter backends though.

Reply:
Let's call it `filter[all-the-things]` ;-)

I agree with your reasoning for calling it something that doesn't imply every field. How about `filter[fields]` or `filter[search_fields]` as the default?

I'm OK with documenting to just use `SearchFilter` and set `search_param`. I just needed a way to make sure when validating filter field names that the one we use for keyword searching doesn't come up as an error. I can just as easily change the code to look for DRF's settings.REST_FRAMEWORK['SEARCH_PARAM'].

I'll drop the code and document example usage of SearchFilter.